File: getting-started-mocks/03-make-failure-boring/node/notifier/notifier.test.js
import assert from 'node:assert/strict'
import test from 'node:test'

import { CarrierUnavailableError, TimeoutError } from './errors.js'
import { Notifier, maxAttempts } from './notifier.js'

// noSleep is the whole reason the sleep seam exists. The retry tests below
// finish instantly instead of waiting for real backoff.
const noSleep = async () => {}

test('creates a notification when the package is delayed', async () => {
  const delayed = async () => 'delayed'
  const notifier = new Notifier(delayed, noSleep)

  assert.equal(await notifier.notify('TRACK-123'), 'Package TRACK-123 is delayed')
})

test('creates no notification when the package is delivered', async () => {
  const delivered = async () => 'delivered'
  const notifier = new Notifier(delivered, noSleep)

  assert.equal(await notifier.notify('TRACK-123'), null)
})

test('reports that the carrier is unavailable', async () => {
  const unavailable = async () => {
    throw new CarrierUnavailableError('carrier returned 500')
  }
  const notifier = new Notifier(unavailable, noSleep)

  await assert.rejects(() => notifier.notify('TRACK-123'), CarrierUnavailableError)
})

test('reports a timeout', async () => {
  const timeout = async () => {
    throw new TimeoutError('carrier did not answer in time')
  }
  const notifier = new Notifier(timeout, noSleep)

  await assert.rejects(() => notifier.notify('TRACK-123'), TimeoutError)
})

test('ignores an unknown status', async () => {
  for (const status of ['', 'in_transit', 'held_at_customs']) {
    const answer = async () => status
    const notifier = new Notifier(answer, noSleep)

    assert.equal(await notifier.notify('TRACK-123'), null, `status ${status}`)
  }
})

test('retries until the carrier answers', async () => {
  let calls = 0
  const flaky = async () => {
    calls++
    if (calls === 1) {
      throw new CarrierUnavailableError('carrier returned 500')
    }
    return 'delayed'
  }
  let sleeps = 0
  const notifier = new Notifier(flaky, async () => {
    sleeps++
  })

  assert.equal(await notifier.notify('TRACK-123'), 'Package TRACK-123 is delayed')
  assert.equal(calls, 2)
  assert.equal(sleeps, 1)
})

test('gives up after the budget', async () => {
  let calls = 0
  const broken = async () => {
    calls++
    throw new CarrierUnavailableError('carrier returned 500')
  }
  const notifier = new Notifier(broken, noSleep)

  await assert.rejects(() => notifier.notify('TRACK-123'), CarrierUnavailableError)
  assert.equal(calls, maxAttempts)
})

File: getting-started-mocks/03-make-failure-boring/go/notifier/notifier_test.go
package notifier

import (
	"context"
	"errors"
	"testing"
	"time"
)

// noSleep is the whole reason the sleep seam exists. The retry tests below
// finish instantly instead of waiting for real backoff.
func noSleep(time.Duration) {}

func TestNotifyDelayedPackage(t *testing.T) {
	delayed := func(string) (string, error) { return "delayed", nil }
	n := New(delayed, noSleep)

	message, err := n.Notify("TRACK-123")
	if err != nil {
		t.Fatal(err)
	}
	if message != "Package TRACK-123 is delayed" {
		t.Fatalf("unexpected message: %q", message)
	}
}

func TestNotifyDeliveredPackage(t *testing.T) {
	delivered := func(string) (string, error) { return "delivered", nil }
	n := New(delivered, noSleep)

	message, err := n.Notify("TRACK-123")
	if err != nil {
		t.Fatal(err)
	}
	if message != "" {
		t.Fatalf("expected no message, got %q", message)
	}
}

func TestNotifyReportsCarrierUnavailable(t *testing.T) {
	unavailable := func(string) (string, error) { return "", ErrCarrierUnavailable }
	n := New(unavailable, noSleep)

	message, err := n.Notify("TRACK-123")
	if !errors.Is(err, ErrCarrierUnavailable) {
		t.Fatalf("expected a carrier-unavailable error, got %v", err)
	}
	if message != "" {
		t.Fatalf("expected no message, got %q", message)
	}
}

func TestNotifyReportsTimeout(t *testing.T) {
	timeout := func(string) (string, error) { return "", context.DeadlineExceeded }
	n := New(timeout, noSleep)

	if _, err := n.Notify("TRACK-123"); !errors.Is(err, context.DeadlineExceeded) {
		t.Fatalf("expected a timeout, got %v", err)
	}
}

func TestNotifyIgnoresAnUnknownStatus(t *testing.T) {
	for _, status := range []string{"", "in_transit", "held_at_customs"} {
		answer := func(string) (string, error) { return status, nil }
		n := New(answer, noSleep)

		message, err := n.Notify("TRACK-123")
		if err != nil {
			t.Fatalf("status %q: %v", status, err)
		}
		if message != "" {
			t.Fatalf("status %q: expected no message, got %q", status, message)
		}
	}
}

func TestNotifyRetriesUntilTheCarrierAnswers(t *testing.T) {
	calls := 0
	flaky := func(string) (string, error) {
		calls++
		if calls == 1 {
			return "", ErrCarrierUnavailable
		}
		return "delayed", nil
	}
	sleeps := 0
	n := New(flaky, func(time.Duration) { sleeps++ })

	message, err := n.Notify("TRACK-123")
	if err != nil {
		t.Fatal(err)
	}
	if message != "Package TRACK-123 is delayed" {
		t.Fatalf("unexpected message: %q", message)
	}
	if calls != 2 {
		t.Fatalf("expected 2 carrier calls, got %d", calls)
	}
	if sleeps != 1 {
		t.Fatalf("expected 1 backoff between attempts, got %d", sleeps)
	}
}

func TestNotifyGivesUpAfterTheBudget(t *testing.T) {
	calls := 0
	broken := func(string) (string, error) {
		calls++
		return "", ErrCarrierUnavailable
	}
	n := New(broken, noSleep)

	if _, err := n.Notify("TRACK-123"); !errors.Is(err, ErrCarrierUnavailable) {
		t.Fatalf("expected a carrier-unavailable error, got %v", err)
	}
	if calls != maxAttempts {
		t.Fatalf("expected %d attempts, got %d", maxAttempts, calls)
	}
}
